src/lib/transition_handler/breadcrumb_builder.ml

```diff
@@ -90,24 +90,42 @@ module Make (Inputs : Inputs.S) :
                            breadcrumb_if_present ()
                          in
                          new_breadcrumb)
-                  | Error (`Fatal_error exn) ->
-                      Deferred.return (Or_error.of_exn exn)
-                  | Error (`Invalid_staged_ledger_hash error) ->
-                      let%map () =
-                        Trust_system.record_envelope_sender trust_system logger
-                          sender
-                          ( Trust_system.Actions.Gossiped_invalid_transition
-                          , Some ("invalid staged ledger hash", []) )
+                  | Error err -> (
+                      (* propagate bans through subtree *)
+                      let subtrees = Rose_tree.flatten subtree in
```

Comment:
`subtree_nodes` might be a better name (`subtrees` to me implies the type should be `'a Rose_tree.t list`, not `'a list`).